src/binary-reader.cc

```diff
@@ -3037,6 +3034,10 @@ Result BinaryReader::ReadModule(const ReadModuleOptions& options) {
   // in case the code section was omitted.
   ERROR_UNLESS(num_function_signatures_ == num_function_bodies_,
                "function signature count != function body count");
+  // This must be checked at the end, in case the data section was omitted.
+  ERROR_UNLESS(
+      data_count_ == kInvalidIndex || data_count_ == num_data_segments_,
+      "data segment count does not equal count in DataCount section");
```

Comment:
Ok, if we must have a separate check can we change the wording a little so we can tell which error is being emitted.

How about changing this one to be something like: "Non-zero DataCount section, but not data segments found", and maybe surround this check with `if (num_data_segments_ == 0)` since that is the only time we need to check for this condition.